packages/react/tooltip/src/Tooltip.tsx

```diff
@@ -330,16 +366,109 @@ interface TooltipContentProps extends TooltipContentImplProps {
+          <TooltipContentImpl side={side} {...contentProps} ref={forwardedRef} />
+        ) : (
+          <TooltipContentHoverable side={side} {...contentProps} ref={forwardedRef} />
+        )}
       </Presence>
     );
   }
 );
 
+type Point = { x: number; y: number };
+type Polygon = Point[];
+
+type TooltipContentHoverableElement = TooltipContentImplElement;
+interface TooltipContentHoverableProps extends TooltipContentImplProps {}
+
+const TooltipContentHoverable = React.forwardRef<
+  TooltipContentHoverableElement,
+  TooltipContentHoverableProps
+>((props: ScopedProps<TooltipContentHoverableProps>, forwardedRef) => {
+  const context = useTooltipContext(TRIGGER_NAME, props.__scopeTooltip);
+  const providerContext = useTooltipProviderContext(TOOLTIP_NAME, props.__scopeTooltip);
+  const ref = React.useRef<TooltipContentHoverableElement>(null);
+  const composedRefs = useComposedRefs(forwardedRef, ref);
+  const [pointerGraceArea, setPointerGraceArea] = React.useState<Polygon | null>(null);
+  const pointerGraceTimerRef = React.useRef(0);
+
+  const { onClose } = context;
+  const content = ref.current;
+  const trigger = context.trigger;
+
+  const handleRemoveGraceArea = React.useCallback(() => {
+    setPointerGraceArea(null);
+    removeDebugArea();
+    window.clearTimeout(pointerGraceTimerRef.current);
+    providerContext.isPointerInTransitRef.current = false;
+  }, [providerContext.isPointerInTransitRef]);
+
+  const handleCreateGraceArea = React.useCallback(
+    (event: PointerEvent, target: HTMLElement) => {
+      const pointerExitPosition = { x: event.clientX, y: event.clientY };
+      const targetPoints = getPointsFromRect(target.getBoundingClientRect());
+      const graceArea = getHull([pointerExitPosition, ...targetPoints]);
+      setPointerGraceArea(graceArea);
+      createDebugArea(graceArea);
+
+      pointerGraceTimerRef.current = window.setTimeout(() => {
+        onClose();
+        handleRemoveGraceArea();
+      }, 500);
+
+      providerContext.isPointerInTransitRef.current = true;
+    },
+    [onClose, handleRemoveGraceArea, providerContext.isPointerInTransitRef]
+  );
+
+  React.useEffect(() => {
+    return () => handleRemoveGraceArea();
+  }, [handleRemoveGraceArea]);
+
+  React.useEffect(() => {
+    if (trigger && content) {
+      const handleTriggerLeave = (event: PointerEvent) => handleCreateGraceArea(event, content);
+      const handleContentLeave = (event: PointerEvent) => handleCreateGraceArea(event, trigger);
+
+      trigger.addEventListener('pointerleave', handleTriggerLeave);
+      content.addEventListener('pointerleave', handleContentLeave);
+      return () => {
+        trigger.removeEventListener('pointerleave', handleTriggerLeave);
+        content.removeEventListener('pointerleave', handleContentLeave);
+      };
+    }
+  }, [trigger, content, context.trigger, handleCreateGraceArea, handleRemoveGraceArea]);
+
+  React.useEffect(() => {
+    if (pointerGraceArea) {
```

Comment:
Is the timer for that reason, so you can clear the area after some time, and make sure we're not tracking the movements anymore?

Reply:
Presence of a `pointerGraceArea` determines whether we need these document handlers or not, so it's not specifically related to the timer. It's more that when we remove the area at any time it also cleans up the event handlers without having to call `removeEventListener` in multiple places.